copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+ * </p>
+ */
+public class KafkaConfigStorage {
+    private static final Logger log = LoggerFactory.getLogger(KafkaConfigStorage.class);
+
+    public static final String CONFIG_TOPIC_CONFIG = "config.storage.topic";
+
+    public static final String ROOT_KEY = "root";
+    public static final String CONNECTOR_PREFIX = "connector-";
+    public static String CONNECTOR_KEY(String connectorName) {
+        return CONNECTOR_PREFIX + connectorName;
+    }
+    public static final String TASK_PREFIX = "task-";
+    public static String TASK_KEY(ConnectorTaskId taskId) {
+        return TASK_PREFIX + taskId.connector() + "-" + taskId.task();
+    }
+
+    // Note that while using real serialization for values as we have here, but ad hoc string serialization for keys,
+    // isn't ideal, we use this approach because it avoids any potential problems with schema evolution or
+    // converter/serializer changes causing keys to change. We need to absolutely ensure that the keys remain precisely
+    // the same.
+    public static final Schema ROOT_V0 = SchemaBuilder.struct()
+            .field("connectors", SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.INT32_SCHEMA).build())
+            .build();
+    public static final Schema CONNECTOR_CONFIGURATION_V0 = SchemaBuilder.struct()
+            .field("properties", SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.OPTIONAL_STRING_SCHEMA))
+            .build();
+    public static final Schema TASK_CONFIGURATION_V0 = CONNECTOR_CONFIGURATION_V0;
+
+    private static final long READ_TO_END_TIMEOUT_MS = 30000;
+
+    private final Object lock;
+    private boolean starting;
+    private final Converter converter;
+    private final Callback<String> connectorConfigCallback;
+    private final Callback<List<ConnectorTaskId>> tasksConfigCallback;
+    private String topic;
+    // Data is passed to the log already serialized. We use a converter to handle translating to/from generic Copycat
+    // format to serialized form
+    private KafkaBasedLog<String, byte[]> configLog;
+    // Root config: connector -> # of tasks
+    private Map<String, Integer> rootConfig = new HashMap<>();
+    // Connector and task configs: name or id -> config map
+    private Map<String, Map<String, String>> connectorConfigs = new HashMap<>();
+    private Map<ConnectorTaskId, Map<String, String>> taskConfigs = new HashMap<>();
+
+    private Map<ConnectorTaskId, Map<String, String>> deferredTaskUpdates = new HashMap<>();
+
+
+    public KafkaConfigStorage(Converter converter, Callback<String> connectorConfigCallback, Callback<List<ConnectorTaskId>> tasksConfigCallback) {
+        this.lock = new Object();
+        this.starting = false;
+        this.converter = converter;
+        this.connectorConfigCallback = connectorConfigCallback;
+        this.tasksConfigCallback = tasksConfigCallback;
+    }
+
+    public void configure(Map<String, ?> configs) {
+        if (configs.get(CONFIG_TOPIC_CONFIG) == null)
```

Comment:
Interesting idea. We might actually be able to do this. The root config only encodes ephemeral data too -- the only data that really *must* be persistent even in the face of all workers dying are the connector configs. I actually dislike the name "root" config but couldn't come up with another name for it. It just keeps track of the number of currently active tasks for each connector. That information can easily be regenerated by re-requesting task configs from a connector.

Separating these two might simplify the coordination a bit. One piece I realized I missed in this patch is that we need to expose the offset up to which we've applied updates so this information can be included when the worker runs JoinGroup/Sync steps to do assignment. This allows us to not include all the configs in the group management protocol itself. But exposing that info is tricky given the current setup because we really have 2 offsets -- the offset of the last connector config update we applied and the offset of the last root key update we've applied.

The obvious drawback is that we'll have to use 2 topics and the code is probably a bit more verbose and a bit redundant across the two classes we'll now have.